File: implementations/dotnet/pcf-dcp/src/Pcf.Dcp/PcfDcpException.cs
using System;

namespace Pcf.Dcp;

/// <summary>Discriminant identifying which kind of <see cref="PcfDcpException"/> occurred.</summary>
public enum PcfDcpErrorKind
{
    /// <summary>The arena did not begin with the <c>"PDCP"</c> magic.</summary>
    BadDcpMagic,
    /// <summary>The arena's profile major version is not implemented.</summary>
    UnsupportedProfileMajor,
    /// <summary>A Fragment Entry carried an unsupported kind (HOLE/REF/unknown).</summary>
    BadFragmentKind,
    /// <summary>An extent range escapes <c>[0, arena_used)</c>.</summary>
    OffsetOutOfRange,
    /// <summary>Reconstructed length did not match <c>used_bytes</c>, or a hash failed.</summary>
    LengthMismatch,
    /// <summary>A stored hash (inner table_hash or data_hash) did not verify.</summary>
    HashMismatch,
    /// <summary>No inner (or top-level) partition with the requested uid.</summary>
    NotFound,
    /// <summary>A uid is used by more than one partition file-wide.</summary>
    DuplicateUid,
    /// <summary>An inner partition is itself a DCP container (nesting forbidden).</summary>
    NestedContainer,
    /// <summary>A partition uid is the PCF NIL uid.</summary>
    NilUid,
    /// <summary>A partition type is the PCF reserved type <c>0x00000000</c>.</summary>
    ReservedType,
    /// <summary>A top-level partition expected to be a DCP container is not one.</summary>
    NotADcpContainer,
    /// <summary>A logical edit addressed a position beyond the partition's content.</summary>
    PositionOutOfRange,
}

/// <summary>All ways a PCF-DCP operation can fail.</summary>
public sealed class PcfDcpException : Exception
{
    /// <summary>The kind of failure.</summary>
    public PcfDcpErrorKind Kind { get; }

    private PcfDcpException(PcfDcpErrorKind kind, string message) : base(message)
    {
        Kind = kind;
    }

    internal static PcfDcpException BadDcpMagic() =>
        new PcfDcpException(PcfDcpErrorKind.BadDcpMagic, "arena does not begin with \"PDCP\" magic");

    internal static PcfDcpException UnsupportedProfileMajor(int v) =>
        new PcfDcpException(PcfDcpErrorKind.UnsupportedProfileMajor,
            $"unsupported PCF-DCP profile major version {v}");

    internal static PcfDcpException BadFragmentKind(int k) =>
        new PcfDcpException(PcfDcpErrorKind.BadFragmentKind, $"unsupported fragment kind {k}");

    internal static PcfDcpException OffsetOutOfRange() =>
        new PcfDcpException(PcfDcpErrorKind.OffsetOutOfRange, "extent range escapes the arena");

    internal static PcfDcpException LengthMismatch(long expected, long got) =>
        new PcfDcpException(PcfDcpErrorKind.LengthMismatch,
            $"logical length mismatch: expected {expected}, got {got}");

    internal static PcfDcpException HashMismatch() =>
        new PcfDcpException(PcfDcpErrorKind.HashMismatch, "stored hash does not verify");

    internal static PcfDcpException NotFound() =>
        new PcfDcpException(PcfDcpErrorKind.NotFound, "no partition with that uid");

    internal static PcfDcpException DuplicateUid() =>
        new PcfDcpException(PcfDcpErrorKind.DuplicateUid, "uid is not unique file-wide");

    internal static PcfDcpException NestedContainer() =>
        new PcfDcpException(PcfDcpErrorKind.NestedContainer,
            "an inner partition may not be a DCP container");

    internal static PcfDcpException NilUid() =>
        new PcfDcpException(PcfDcpErrorKind.NilUid, "uid is the NIL uid");

    internal static PcfDcpException ReservedType() =>
        new PcfDcpException(PcfDcpErrorKind.ReservedType,
            "partition type is the reserved type 0x00000000");

    internal static PcfDcpException NotADcpContainer() =>
        new PcfDcpException(PcfDcpErrorKind.NotADcpContainer, "partition is not a DCP container");

    internal static PcfDcpException PositionOutOfRange() =>
        new PcfDcpException(PcfDcpErrorKind.PositionOutOfRange,
            "logical position is past end of content");
}

File: implementations/php/pcf-dcp/src/PcfDcpException.php
<?php

declare(strict_types=1);

namespace Kduma\PCFDCP;

/**
 * The single exception type raised by this library. The {@see ErrorKind}
 * carried in {@see PcfDcpException::$kind} identifies the precise failure.
 */
final class PcfDcpException extends \RuntimeException
{
    public function __construct(
        public readonly ErrorKind $kind,
        string $message,
        ?\Throwable $previous = null,
    ) {
        parent::__construct($message, 0, $previous);
    }

    public static function badDcpMagic(): self
    {
        return new self(ErrorKind::BadDcpMagic, 'arena does not begin with "PDCP" magic');
    }

    public static function unsupportedProfileMajor(int $v): self
    {
        return new self(ErrorKind::UnsupportedProfileMajor, "unsupported PCF-DCP profile major version {$v}");
    }

    public static function badFragmentKind(int $k): self
    {
        return new self(ErrorKind::BadFragmentKind, "unsupported fragment kind {$k}");
    }

    public static function offsetOutOfRange(): self
    {
        return new self(ErrorKind::OffsetOutOfRange, 'extent range escapes the arena');
    }

    public static function lengthMismatch(int $expected, int $got): self
    {
        return new self(ErrorKind::LengthMismatch, "logical length mismatch: expected {$expected}, got {$got}");
    }

    public static function hashMismatch(): self
    {
        return new self(ErrorKind::HashMismatch, 'stored hash does not verify');
    }

    public static function notFound(): self
    {
        return new self(ErrorKind::NotFound, 'no partition with that uid');
    }

    public static function duplicateUid(): self
    {
        return new self(ErrorKind::DuplicateUid, 'uid is not unique file-wide');
    }

    public static function nestedContainer(): self
    {
        return new self(ErrorKind::NestedContainer, 'an inner partition may not be a DCP container');
    }

    public static function nilUid(): self
    {
        return new self(ErrorKind::NilUid, 'uid is the NIL uid');
    }

    public static function reservedType(): self
    {
        return new self(ErrorKind::ReservedType, 'partition type is the reserved type 0x00000000');
    }

    public static function notADcpContainer(): self
    {
        return new self(ErrorKind::NotADcpContainer, 'partition is not a DCP container');
    }

    public static function positionOutOfRange(): self
    {
        return new self(ErrorKind::PositionOutOfRange, 'logical position is past end of content');
    }
}
